[BCFR-900] Add id column as PRIMARY KEY for evm.logs & evm.log_poller_blocks

reductionista opened the pull request:
The query for pruning expired logs with a max limit set was taking longer than it should. This was  due in part to needing to join on an awkward combination combination of columns due to their being no single primary key.

- Add id column as PRIMARY KEY for evm.logs & evm.log_poller_blocks
- Join on id column instead of previous primary keys
- Replace all SELECT *'s with helper functions for selecting all columns
- Refactor nestedBlockQuery into withConfs, and make a bit more use of it## Motivation

- While adding the id column, we can't just remove the old primary key because the index on it was helping to accelerate some queries.
   Instead of just resurrecting it as-is, I took the opportunity to clean up several of the indices on the logs & blocks table. Some indexed columns (eg created_at) were never actually being used,
   while others were not ordered in the most optimal way for accelerating the queries we have. Also, at least one of them was redundant with the primary key just in a different order.

- After testing the performance of the `DeleteExpiredLogs` query, it was found that using the id column makes a huge difference to the part of the query involving actual expired logs. But it makes no difference to the part which finds logs not matching any filter--which is the main thing keeping the query from being faster. To address this, the pruning of unmatched logs has been separated off into its own method `PruneExpiredLogs` which only every 20th time `PruneExpiredLogs` succeeds.  Both of the queries were faster on their own than together anyway, and there is no reason to prune logs corresponding to old jobs as frequently as expired logs for present jobs since those are not actively growing and are only a long-term issue.

- This also includes an update to the `DeleteBlocksBefore` query which drastically improves its performance. Instead of using a `LIMIT` clause, which complicates the `DELETE` query, it uses an iterative range filter by block number, starting at `MIN(block_number)` and continuing upwards in windows of `limit` blocks until it has either removed everything it can or hits the limit.  (Hitting the limit on # of blocks deleted can take more than 1 iteration because we don't necessarily save every block.)

[BCFR-900](https://smartcontract-it.atlassian.net/browse/BCFR-900)

[BCFR-900]: https://smartcontract-it.atlassian.net/browse/BCFR-900?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

reductionista committed:
Add id column as PRIMARY KEY for evm.logs & evm.log_poller_blocks

Also:
   - Add UNIQUE INDEXes to replace previous primary keys (still necessary, both for
     optimizing queries and for enforcing uniqueness constraints)
   - Replace all SELECT *'s with helper functions for selecting all columns
   - Refactor nestedBlockQuery into withConfs, and make a bit more use of it

reductionista committed:
Clean up db indexes

Some of the columns in these indexes (such as created_at) are no longer used.
Others were not optimized for the queries we need.

reductionista committed:
Fix 2 unrelated bugs I noticed

reductionista committed:
Update ExpiredLogs query to use id column

reductionista committed:
Update test for fromBlock >= :block_number

Previously it was using fromBlock > :block_number which is inconsistent with the other fromBlocks in queries

reductionista committed:
Increase staggering of initial pruning runs

reductionista committed:
Update go version 1.22.5 -> 1.22.7

reductionista committed:
Update DeleteBlocksBefore query to use block_number index instead of LIMIT

reductionista committed:
Split off SelectUnmatchedLogs from DeleteExpiredLogs

reductionista committed:
Rename Id -> ID for linter

mateusz-sekara commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -635,31 +635,50 @@ func (lp *logPoller) backgroundWorkerRun() {
 				// Tick faster when cleanup can't keep up with the pace of new logs
-				logPruneTick = time.After(utils.WithJitter(lp.pollPeriod * 241))
+				logPruneTick = time.After(timeutil.JitterPct(0.1).Apply(logPruneShortInterval))
+			} else if successfulExpiredLogPrunes == 20 {
```
There is a risk of never entering this when escape velocity exceeds pruning speed. I can imagine an edge scenario in which we cannot keep up with pruning expired logs, because there is an insane number of events produced by the chain - it doesn't have to be the traffic on a chain, it could be caused by the LogPoller misconfig, e.g. setting too long poll period on a fast chain.

How about making it simpler and running `PruneUnmatched` after successful 20 runs no matter whether all was removed or not
```go
	logPruneTick = time.After(timeutil.JitterPct(0.1).Apply(logPruneInterval))
	if allRemoved, err := lp.PruneExpiredLogs(ctx); err != nil {
		lp.lggr.Errorw("Unable to prune expired logs", "err", err)
	} else {
		successfulExpiredLogPrunes++

		if !allRemoved {
			// Tick faster when cleanup can't keep up with the pace of new logs
			logPruneTick = time.After(timeutil.JitterPct(0.1).Apply(logPruneShortInterval))
		} else if successfulExpiredLogPrunes == 20 {
			// Only prune unmatched logs if we've successfully pruned all expired logs at least 20 times
			// since the last time unmatched logs were pruned
			if allRemoved, err := lp.PruneUnmatchedLogs(ctx); err != nil {
				lp.lggr.Errorw("Unable to prune unmatched logs", "err", err)
			} else if !allRemoved {
				logPruneTick = time.After(timeutil.JitterPct(0.1).Apply(logPruneShortInterval))
			} else {
				successfulExpiredLogPrunes = 0
			}
		}
	}	
```

reductionista replied:
Until recently, we never even pruned old logs that didn't match a filter, so it seems like it's the kind of operation that should have the lowest priority compared to anything else. If it's already so overloaded that it's unable to keep up with pruning expired logs then running it risks causing other queries to time out... making the problem worse.

So my thinking is: it seems safer to not do it unless we're sure the system is functioning smoothly and can handle a query that may lock a lot of the table and take longer than most queries.  Even if it had to wait a year for such a moment, that's probably fine as the accumulation of logs from old jobs or lanes is only a very long term problem. Meanwhile, if the system is not keeping up with pruning expired logs that's going to need to be fixed with some sort of configuration change (eg decreasing page size, or increasing cpu limits for the container) soon anyway... pruning the unmatched logs seems like a lesser concern.

What do you think?

mateusz-sekara replied:
Yeah, you are right. If we cannot keep up with deleting we should not bother ourselves with removing some super stale data.

mateusz-sekara commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -635,31 +635,50 @@ func (lp *logPoller) backgroundWorkerRun() {
+				if allRemoved, err := lp.PruneUnmatchedLogs(ctx); err != nil {
+					lp.lggr.Errorw("Unable to prune unmatched logs", "err", err)
+				} else if !allRemoved {
+					logPruneTick = time.After(timeutil.JitterPct(0.1).Apply(logPruneShortInterval))
```
nit, we do a lot of these `time.After(timeutil.JitterPct(0.1).Apply(interval))`. How about extracting that to some helper like
```go
func tickWithJitter(interval time.Duration) <-chan time.Time {
	return time.After(timeutil.JitterPct(0.1).Apply(interval))
}
```

mateusz-sekara commented on core/chains/evm/logpoller/observability.go:
```diff
@@ -136,6 +136,18 @@ func (o *ObservedORM) DeleteLogsAndBlocksAfter(ctx context.Context, start int64)
 	})
 }
 
+func (o *ObservedORM) DeleteLogsByRowID(ctx context.Context, rowIDs []uint64) (int64, error) {
```
❤️

mateusz-sekara commented on core/store/migrate/migrations/0254_log_poller_primary_keys.sql:
```diff
@@ -0,0 +1,24 @@
+CREATE UNIQUE INDEX idx_logs_chain_block_logindex ON evm.logs (evm_chain_id, block_number, log_index);
+ALTER TABLE evm.log_poller_blocks DROP CONSTRAINT log_poller_blocks_pkey;
+ALTER TABLE evm.log_poller_blocks ADD COLUMN id BIGINT PRIMARY KEY GENERATED ALWAYS AS IDENTITY;
+DROP INDEX IF EXISTS evm.idx_evm_log_poller_blocks_order_by_block;
```
Is it because we need to make it unique?

mateusz-sekara commented on core/store/migrate/migrations/0254_log_poller_primary_keys.sql:
```diff
@@ -0,0 +1,24 @@
+DROP INDEX IF EXISTS evm.idx_evm_log_poller_blocks_block_number_evm_chain_id;
+CREATE UNIQUE INDEX idx_log_poller_blocks_chain_block ON evm.log_poller_blocks (evm_chain_id, block_number DESC);
+DROP INDEX IF EXISTS evm.idx_evm_logs_ordered_by_block_and_created_at;
+CREATE INDEX idx_logs_chain_address_event_block_logindex ON evm.logs (evm_chain_id, address, event_sig, block_number, log_index);
```
Where do we rely on that index?

reductionista replied:
I think most of the queries of the logs table should be able to use it. For example, the query for `SelectLogs()` is:

```
    query := logsQuery(`
        WHERE evm_chain_id = :evm_chain_id
        AND address = :address
        AND event_sig = :event_sig
        AND block_number >= :start_block
        AND block_number <= :end_block
        ORDER BY block_number, log_index`)
```

Like most of the logs queries, it filtesr on chain id, address, and event sig and then sorts on block_number, log_index (the order in which logs appear on chain).

reductionista replied:
I got rid of a similar index we had for `created_at`, which was used at one point for ordering instead of `block_number, log_index` but we no longer use `created_at` for anything since it doesn't reliably indicate the order in which the events occurred on chain.

mateusz-sekara replied:
Ahh, right, makes sense

mateusz-sekara replied:
Yeah, that makes sense, but isn't that already covered by one of the existing indexes?
```sql
create index evm_logs_idx
    on evm.logs (evm_chain_id, block_number, address, event_sig);

create index evm_logs_by_timestamp
    on evm.logs (evm_chain_id, address, event_sig, block_timestamp, block_number);
```

Usually, there are not many matching logs within a single block, so probably gain from adding log_index to the index won't be that high, right? Postgres should be able to sort those small datasets efficiently by using the log_index in memory. 

I know you are replacing the `created_at` index one with the new one. Therefore, there should not be any impact on the db size because of that - there were some teams in the past very unhappy with us adding more indexes and growing db :D  I'm fine with having that exploration in a separate PR/ticket.

reductionista replied:
> postgres should be able to sort those small datasets efficiently by using the log_index in memory.

That's true, I'd be fine with omitting the `log_index` column as it's not adding much.  On the other hand, I wouldn't expect that including it would increase the size of that index much.

> isn't that already covered by one of the existing indexes?

I figured it probably can't use either of those for this, but admittedly I haven't tested this.

My general understanding of how these BTREE indices work based on what I've seen in previous plans is that you have to have all of the columns you're filtering on first, and then the sort columns after.  Assuming that is how BTREE's work (I'm not 100% sure about that), the `evm_chain_id, block_number, address, event_sig` index will help for block range queries where you're looking for events of specific types within a particular block range.  While this one would help with ordering the results for events of a specific type.

The one with `block_timestamp` in it seems more likely to work. If it had a way of knowing for sure that block numbers and block timestamps will be ordered in the same way, then it should be able to use it... but I don't think it can know that so my assumption would be that it  probably won't use it.

reductionista replied:
Side thought:  from a database perspective, `block_timestamp` really ought to have been a column of the `blocks` table rather than the `logs` table. We could have easily joined those on block_number to search by block_timestamp without much cost. I guess the reason why it's in the `logs` table is mostly so that it can be returned by the queries that return `[]Log` and to keep the `logs` schema in line with that data struct. Not sure if this decision might be something to revisit in the future.

reductionista replied:
> the evm_chain_id, block_number, address, event_sig index will help for block range queries where you're looking for events of specific types within a particular block range

After writing this I realized this is not right. If this were the only use case, then we'd be able to get rid of the  `evm_chain_id, block_number, address, event_sig` index, and just use `evm_chain_id, address, event_sig, block_number, log_index` for everything... because the SELECT queries where we're looking for events in a particular block range also filter on event_sig and address... so it can just use the `evm_chain_id, address, event_sig, block_number, log_index` index.

Unfortunately, I just looked through the code and I do see one query that filters on block range but not address or event sig. It's where we delete all of the logs and blocks after a particular block due to a re-org:
```
DELETE FROM evm.logs
                         WHERE evm_chain_id = $1
                          AND block_number >= $2
                          AND block_number <= (SELECT MAX(block_number) FROM evm.logs WHERE evm_chain_id = $1)
 ```
So I think we can get rid of the `address, event_sig` columns at the end of this index to save some space.

mateusz-sekara commented on core/chains/evm/logpoller/parser.go:
```diff
@@ -26,6 +27,10 @@ const (
 
 var (
 	ErrUnexpectedCursorFormat = errors.New("unexpected cursor format")
+	logsFields                = [...]string{"evm_chain_id", "log_index", "block_hash", "block_number",
```
I guess `created_at` is not used anywhere in the LogPoller, probably you can drop that from select for both block and logs, right?

reductionista replied:
This field is returned to all of the products using LogPoller in the `logpoller.Log` response from all queries:
```
// Log represents an EVM log.
type Log struct {
    EvmChainId     *big.Big
    LogIndex       int64
    BlockHash      common.Hash
    BlockNumber    int64
    BlockTimestamp time.Time
    Topics         pq.ByteaArray
    EventSig       common.Hash
    Address        common.Address
    TxHash         common.Hash
    Data           []byte
    CreatedAt      time.Time
}

// LogPollerBlock represents an unfinalized block
// used for reorg detection when polling.
type LogPollerBlock struct {
    EvmChainId *big.Big
    BlockHash  common.Hash
    // Note geth uses int64 internally https://github.com/ethereum/go-ethereum/blob/f66f1a16b3c480d3a43ac7e8a09ab3e362e96ae4/eth/filters/api.go#L340
    BlockNumber          int64
    BlockTimestamp       time.Time
    FinalizedBlockNumber int64
    CreatedAt            time.Time
}
```
I don't know if any of them actually care about it, but we can't just return the `CreatedAt` field as 0 for everything as that would be seen as a bug.  I think there's a good case to be made that we could probably just remove this field entirely, but because that's a user-facing change I wouldn't want to do that without checking with product teams and surveying the codebase. (Also, I'd probably want to either remove it for both `Logs` and `LogPollerBlock` or neither, to keep things more consistent.)  I'd say that's out of scope for this PR, but I can create a JIRA for it.

mateusz-sekara replied:
Makes sense, creating a JIRA item in the backlog to rethink this field is good enough ;)

mateusz-sekara commented on core/store/migrate/migrations/0254_log_poller_primary_keys.sql:
```diff
@@ -0,0 +1,24 @@
+DROP INDEX IF EXISTS evm.idx_evm_log_poller_blocks_order_by_block;
+DROP INDEX IF EXISTS evm.idx_evm_log_poller_blocks_block_number_evm_chain_id;
+CREATE UNIQUE INDEX idx_log_poller_blocks_chain_block ON evm.log_poller_blocks (evm_chain_id, block_number DESC);
+DROP INDEX IF EXISTS evm.idx_evm_logs_ordered_by_block_and_created_at;
```
I think it would be great to add a comment to every index creation/removal for future generations to easier identify the index with the relevant query ;)

mateusz-sekara commented on core/store/migrate/migrations/0254_log_poller_primary_keys.sql:
```diff
@@ -0,0 +1,24 @@
+-- +goose Up
+
+ALTER TABLE evm.logs DROP CONSTRAINT logs_pkey;
```
nit, but can we group these DDLs by the table? I mean

```sql
ALTER TABLE evm.logs DROP CONSTRAINT logs_pkey;
ALTER TABLE evm.logs ADD COLUMN id BIGINT PRIMARY KEY GENERATED ALWAYS AS IDENTITY;
CREATE UNIQUE INDEX idx_logs_chain_block_logindex ON evm.logs (evm_chain_id, block_number, log_index);
DROP INDEX IF EXISTS evm.idx_evm_logs_ordered_by_block_and_created_at;
CREATE INDEX idx_logs_chain_address_event_block_logindex ON evm.logs (evm_chain_id, address, event_sig, block_number, log_index);

ALTER TABLE evm.log_poller_blocks DROP CONSTRAINT log_poller_blocks_pkey;
ALTER TABLE evm.log_poller_blocks ADD COLUMN id BIGINT PRIMARY KEY GENERATED ALWAYS AS IDENTITY;
DROP INDEX IF EXISTS evm.idx_evm_log_poller_blocks_order_by_block;
DROP INDEX IF EXISTS evm.idx_evm_log_poller_blocks_block_number_evm_chain_id;
CREATE UNIQUE INDEX idx_log_poller_blocks_chain_block ON evm.log_poller_blocks (evm_chain_id, block_number DESC);
```

mateusz-sekara commented on core/store/migrate/migrations/0254_log_poller_primary_keys.sql:
```diff
@@ -0,0 +1,24 @@
+
+ALTER TABLE evm.logs DROP CONSTRAINT logs_pkey;
+ALTER TABLE evm.logs ADD COLUMN id BIGINT PRIMARY KEY GENERATED ALWAYS AS IDENTITY;
+CREATE UNIQUE INDEX idx_logs_chain_block_logindex ON evm.logs (evm_chain_id, block_number, log_index);
```
Do we need it to be an index? I think for most of the cases we query by at least (evm_chain_id, address, event_sig) so this index might not be needed, but you need to verify that.

If it's only about uniqueness then adding constraints will be more lightweight than indexes

reductionista replied:
That's a good point, I'll check on this.

reductionista replied:
After checking on this, I realized why I originally made it an index.

I don't think it's true about a unique constraint being more lightweight. Every unique constraint generates an unique index automatically... the only difference is we pick the name here rather than it picking one automatically.

I mentioned in another comment that we need at least (chain id, block number) as an index, for the DeleteLogsAndBlocksAfter query. But if we add a unique constraint, that would generate an entire second index... taking up twice the disk space. By doing it this way instead, all we have to do is add log_index to the end and then there is no need to have 2 separate indexes... which is what we had before (due to the one generated by the primary key).

In the end I decided it seems best to leave everything as-is, but I added a lot of comments explaining what's going on and separated the tables out as you suggested.

mateusz-sekara commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -248,28 +296,46 @@ func (o *DSORM) SelectLatestLogByEventSigWithConfs(ctx context.Context, eventSig
+
+	// Remove up to limit blocks at a time, until we've reached the limit or removed everything eligible for deletion
+	var deleted, rows int64
+	for limitBlock += (limit - 1); deleted < limit; limitBlock += limit {
```
Hmm, why do you need to define `limitBlock` outside of the loop? Wouldn't that work with just
```go
	for limitBlock := (limit - 1); deleted < limit; limitBlock += limit {
```

reductionista replied:
It has to be outside the loop because it needs to be initialized as `MIN(block_number) + (limit - 1)`, and the first part of that requires a one-time query. You may have missed that it's in this line above:
```
err = o.ds.GetContext(ctx, &limitBlock, `SELECT MIN(block_number) FROM evm.log_poller_blocks`)
```

Some additional context: originally, I was hoping we could just include the `MIN(block_number)` in the `DELETE` query that's inside the loop. But apparently aggregate functions like `MIN` are not allowed in DELETE queries.

mateusz-sekara reviewed (commented):
Great improvements, left some minor comments

reductionista committed:
Add helper function for tickers

reductionista committed:
Stagger initial unmatched logs prune

reductionista committed:
Reorganize sql migration script, adding comments to clarify

Also, adds a missing CREATE INDEX line that was left out of the Goose Down section:
   CREATE INDEX idx_evm_log_poller_blocks_order_by_block ON evm_log_poller_blocks (evm_chain_id, block_number DESC);

And removes an extraneous DROP line that was in the Goose Up section:
  DROP INDEX IF EXISTS evm.idx_logs_chain_address_event_block_logindex;

reductionista committed:
Remove log_index from idx_logs_chain_block_logindex to save some disk space

dhaidashenko commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -248,28 +296,46 @@ func (o *DSORM) SelectLatestLogByEventSigWithConfs(ctx context.Context, eventSig
-       WHERE block_number <= $1 AND evm_chain_id = $2`, end, ubig.New(o.chainID))
+
+	var limitBlock int64
+	err = o.ds.GetContext(ctx, &limitBlock, `SELECT MIN(block_number) FROM evm.log_poller_blocks`)
```
```suggestion
	err = o.ds.GetContext(ctx, &limitBlock, `SELECT MIN(block_number) FROM evm.log_poller_blocks WHERE evm_chain_id = $1`, ubig.New(o.chainID)`)
```

reductionista replied:
Thanks, nice catch!

dhaidashenko commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -248,28 +296,46 @@ func (o *DSORM) SelectLatestLogByEventSigWithConfs(ctx context.Context, eventSig
+	var limitBlock int64
+	err = o.ds.GetContext(ctx, &limitBlock, `SELECT MIN(block_number) FROM evm.log_poller_blocks`)
 	if err != nil {
 		return 0, err
```
nit: Should we handle errNoRows here?

dhaidashenko commented on core/chains/evm/logpoller/orm.go:
```diff
@@ -248,28 +296,46 @@ func (o *DSORM) SelectLatestLogByEventSigWithConfs(ctx context.Context, eventSig
+			return deleted, err
+		}
+
+		deleted += rows
```
nit: we can return if `rows == 0`

reductionista replied:
Not necessarily. There could be a streak of `limit` blocks that we didn't save due to a backfill, followed by a lot of blocks eligible for deletion. In that scenario, we might only end up deleting 1 or 2 blocks when we could have deleted say, 1000 blocks.

I guess the consequences of returning early are not huge (just, not deleting some logs we could have, while still remaining under MaxPageSize. On the other hand, skipping one more query to see if there are more logs to delete doesn't save much... I think I'd prefer to keep it as is.

dhaidashenko commented on core/chains/evm/logpoller/log_poller.go:
```diff
@@ -635,31 +645,52 @@ func (lp *logPoller) backgroundWorkerRun() {
 		case <-logPruneTick:
-			logPruneTick = time.After(utils.WithJitter(lp.pollPeriod * 2401)) // = 7^5 avoids common factors with 1000
+			logPruneTick = tickWithDefaultJitter(logPruneInterval)
 			if allRemoved, err := lp.PruneExpiredLogs(ctx); err != nil {
```
`allRemoved` and the comment to `PruneOldBlocks` are a bit misleading now. As we always remove all eligible blocks in a single method call now.

reductionista replied:
> allRemoved and the comment to PruneOldBlocks are a bit misleading now. 

If that's true then that's a bug... we should never be deleting more than `limit` number of blocks. I think the code is right, but if you check it again and it still looks like we're deleting all eligible blocks each time, let me know and I'll fix it.

dhaidashenko reviewed (changes requested):
One of the queries misses a chainID clause, but otherwise looks good. Great job!

reductionista committed:
add WHERE evm_chain_id and handle sql.ErrNoRows

reductionista committed:
Fix lint

reductionista committed:
Merge branch 'develop' into BCFR-900-log-poller-id-columns
